### PR TITLE
Add message signalling when coercion produces NAs

### PR DESCRIPTION
**Pull request by dgkf**
Trying to coerce a string to a `partial_time` using `as.parttime` will now emit messages when `NA`s are introduced. It will include some light stats about the number of items that were coerced to `NA`s and a sampling of styles that failed to parse. 

For example:

```r
as.parttime(c("2022", "", "2022-02", "Y1234", "2022-W300", "2023", "Y4321", "02-01-01"))
# <partial_time<YMDhms+tz>[8]>
# [1] "2022"    NA        "2022-02" NA        NA        "2023"    NA        NA
# Warning message:
# In vec_cast.partial_time.character(x, pttm, ..., format = format,  :
#   Values could not be parsed (5 of 8 (62.5%)). Examples of unique failing formats:
# 
#     ''           'Y1234'      '2022-W300'  '02-01-01'
#
```

> Note that only `"Y1234"` is printed to represent strings like `"Y1234"` and `"Y4321"` since they both have the same format. This is a little heuristical, but helps to cut down on the noise and hopefully provide something immediately helpful for finding the uniquely problematic data. The first 10 failing formats are displayed with a trailing `...` if more failing formats exist.

In contrast to #32, I think it's better to emit a warning by default since the output can be useful for examining which indices produced `NA`s. However, this can easily be emitted as an error if desired by passing  `on.na = "error"`. For fine control over how the message is surfaced, it will also accept a function, but this is probably unnecessary in most cases. 

@billdenney - I would appreciate your thoughts on whether this is an acceptable first pass at #32.

### CODE REVIEW

File: R/class_partial_time_coercion.R [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
+#'   `function`, it should accept a character vector and return a matrix of
+#'   parttime components. If a `character` it should provide a regular
 #'   exprssion which contains capture groups for each of the parttime
-#'   components.  See \code{?parse_to_parttime_matrix}'s \code{regex} parameter
+#'   components.  See [parse_to_parttime_matrix]'s `regex` parameter
 #'   for more details.
-#' @param ... Additional arguments passed to \code{format} when a function is
+#' @param on.na a `function` used to signal a condition for new `NA` values
+#'   introduced by coercion, a `character` value among `"error"`, `"warning"` or
+#'   `"suppress"` (for silencing messages) or `NULL` equivalent to `"suppress"`.
```
Comment (dgkf) [inline R/class_partial_time_coercion.R:12]: I did some code cleanup, so just for the sake of drawing attention to the important bits, this is the new `on.na` parameter.

File: R/class_partial_time_coercion.R [hunk 2]
```diff
@@ -36,15 +39,11 @@
...
   # spoof a parttime class object for dispatch to prevent recursion since
   # parttime()  function uses as.parttime.matrix
-  vec_cast.partial_time(
-    x,
-    structure(0L, class = "partial_time"),
-    ...,
-    format = format
-  )
+  pttm <- structure(0L, class = "partial_time")
+  vec_cast.partial_time(x, pttm, ..., format = format, on.na = on.na)
```
Comment (dgkf) [inline R/class_partial_time_coercion.R:46]: Minor refactor to cleanup the call emitted with errors and warnings.

File: R/class_partial_time_coercion.R [hunk 3]
```diff
@@ -96,41 +95,46 @@ vec_cast.partial_time.default <- function(x, to, ...) {
...
   }
 
   pttm_mat <- clean_parsed_parttime_matrix(pttm_mat)
-  as.parttime(pttm_mat)
+  res <- as.parttime(pttm_mat)
+
+  # if NAs are introduced during coercsion, emit on.na callback
+  if (any(is.na(res) & !is.na(x))) {
+    on.na(parse_failure_message(x, res))
+  }
```
Comment (dgkf) [inline R/class_partial_time_coercion.R:135]: This is where the message gets emitted, any time a parsing function produces new `NA`s, not found in the input. This just compares the results to the input, so will extend automatically to cdisc-style datetimes once a format is added.

**Comment by billdenney** [comment]
At first look, I think that it covers the goals well.

You went to some more notable effort than I would have for showing the error strings.  As long as that will always give a result, I think that it's better.  (I didn't do a detailed code review of that part, but my concern with that is that it adds complexity which could have issues.)

**Comment by dgkf** [comment]
@billdenney cool - I think things are in good shape, but I think it's better than the quiet coercion that happens now. Please continue to report bugs if you find them!

**Pull request merged**